shoulder/__main__.py

```diff
@@ -21,82 +21,17 @@
 # SOFTWARE.
 
 import os
-import glob
 
-from shoulder.parser.armv8_xml_parser import ArmV8XmlParser
-from shoulder.generator.c_header_generator import CHeaderGenerator
-from shoulder.generator.cxx_header_generator import CxxHeaderGenerator
 from shoulder.config import config
 from shoulder.logger import logger
+from shoulder.parser import *
+from shoulder.generator import *
 from shoulder.filters import *
 
-def printReg(reg):
-    logger.info("NAME:\n    " + reg.name)
-    logger.info("LONG NAME:\n    " + reg.long_name)
-    logger.info("PURPOSE:\n    " + reg.purpose)
-    logger.info("SIZE:\n    " + str(reg.size))
-    logger.info("OFFSET:\n    " + str(reg.offset))
-    logger.info("IS_SYSREG:\n    " + str(reg.is_sysreg))
-    msg = ""
-    for fieldset in reg.fieldsets:
-        msg += "\n" + str(fieldset)
-    logger.info("FIELDSETS:" + msg)
+regs = parse_registers(config.xml_register_dir)
+logger.info("Registers parsed: " + str(len(regs)))
 
-
-# SETUP
-logger.set_log_level("info")
-test_input_dir = "/home/jzepf/Projects/Scapula/ARM-SysReg-v85A-00bet9/SysReg_v85A_xml-00bet9/"
-config.c_prefix = "aarch"
-
-# PARSE
-logger.info("Parsing registers from: " + str(test_input_dir))
-paths = glob.glob(test_input_dir + "*.xml")
-regs = []
-parser = ArmV8XmlParser()
-for path in paths:
-    results = parser.parse_registers(path)
-    if results:
-        regs.append(results[0])
-
-# FILTER
 regs = apply_filters(regs)
+logger.info("Registers after filters: " + str(len(regs)))
 
-# GENERATE
-if not os.path.exists(config.shoulder_output_dir):
-    os.makedirs(config.shoulder_output_dir)
-c_generator = CHeaderGenerator()
-c_generator.generate(regs, (config.shoulder_output_dir + "/c_shoulder.h"))
-
-cxx_generator = CxxHeaderGenerator()
-cxx_generator.generate(regs, (config.shoulder_output_dir + "/cxx_shoulder.h"))
-
-
-
-#  import os
-#  import glob
-#
-#  from shoulder.parser.armv8_xml_parser import ArmV8XmlParser
-#  from shoulder.generator import *
-#  from shoulder.config import config
-#  from shoulder.logger import logger
-#  from shoulder.filters import *
-#
-#  # TODO Import configuration from command line arguments
-#  test_input_dir = "/Users/jared-ais/workspace/thunderlane/ARM_ASL/mra_tools/v8.3/SysReg_v83A_xml-00bet5/"
-#  config.cxx_namespace = "aarch64"
-#
-#  # Parse all input registers
-#  logger.info("Parsing registers from: " + str(test_input_dir))
-#  paths = glob.glob(test_input_dir + "*.xml")
-#  regs = []
-#  parser = ArmV8XmlParser()
-#  for path in paths:
-#      results = parser.parse_registers(path)
-#      if results:
-#          regs.append(results[0])
-#
-#  # Apply filters (shoulder.filters.*)
-#  regs = apply_filters(regs)
-#
-#  # Generate output files (shoulder.generator.*)
-#  generate_all(regs, config.shoulder_output_dir)
+generate_all(regs, config.shoulder_output_dir)
```

Comment:
I updated this to automatically generate everything in the shoulder/generator module that subclasses AbstractGenerator. Each concrete generator that subclasses AbstractGenerator gets a unique output directory under shoulder/output, based on the name of the class. Eventually, I want to provide a mechanism for user to "register" new generators to shoulder from their own projects

Reply:
Yeah, I like this a lot better. It makes the code much cleaner and the output more organized. Someone could register a generator from their own project by importing AbstractGenerator and inheriting it in their generator class which seems like the most intuitive way to go about that.